src/DynamicData.Tests/Utilities/ObservableSpy.cs

```diff
@@ -71,110 +81,131 @@ internal static class ObservableSpy
                                                                       where T : notnull =>
-#if DEBUG
+#if DEBUG || DEBUG_SPY_ALWAYS
         source.Spy(opName, DebugLogger, formatter, showSubs, showTimestamps);
 #else
         source;
 #endif
 
-
     private static string FormatChange<T, TKey>(Func<T, string> formatter, Change<T, TKey> change)
         where T : notnull
         where TKey : notnull =>
-        change.Reason switch
-        {
-            ChangeReason.Update => $"{formatter(change.Current)} [Previous: {formatter(change.Previous.Value)}]",
-            _ => formatter(change.Current),
-        };
+        $"[{change.Reason}] " +
+            change.Reason switch
+            {
+                ChangeReason.Update => $"{formatter(change.Current)} [Previous: {formatter(change.Previous.Value)}]",
+                _ => formatter(change.Current),
+            };
 
     private static string FormatChange<T>(Func<T, string> formatter, Change<T> change)
         where T : notnull =>
-        change.Reason switch
-        {
-            ListChangeReason.AddRange => string.Join(", ", change.Range.Select(n => formatter(n))),
-            ListChangeReason.RemoveRange => string.Join(", ", change.Range.Select(n => formatter(n))),
-            _ => formatter(change.Item.Current),
-        };
+        $"[{change.Reason}] " +
+            change.Reason switch
+            {
+                ListChangeReason.AddRange => FormatRangeChange(formatter, change.Range),
+                ListChangeReason.RemoveRange => FormatRangeChange(formatter, change.Range),
+                _ => FormatItemChange(formatter, change.Item),
+            };
+
+    private static string FormatRangeChange<T>(Func<T, string> formatter, RangeChange<T> range) where T : notnull =>
+        $"({range.Count} Values) "
+            + (range.Index == -1 ? string.Empty : $"[Index: {range.Index}] ")
+            + string.Join(", ", range.Select(n => formatter(n)));
+
+    private static string FormatItemChange<T>(Func<T, string> formatter, ItemChange<T> item) where T : notnull =>
+        formatter(item.Current)
+            + ((item.CurrentIndex, item.PreviousIndex) switch
+            {
+                (-1, -1) => string.Empty,
+                (int i, int j) when i == j || (j == -1) => $" [Index: {i}]",
+                (int i, int j) => $" [Index: {i}, Prev: {j}]",
+            })
+            + (item.Previous switch
+            {
+                { HasValue: true, Value: T val } => $" [Previous: {formatter(val)}]",
+                _ => string.Empty,
+            });
 
     private static Action<string> CreateLogger(Action<string> baseLogger, Func<string> timeStamper, string opName) =>
             msg => baseLogger($"{timeStamper()}[{Environment.CurrentManagedThreadId:X2}] |{opName}| {msg}");
 
-#if DEBUG
-    static void DebugLogger(string str) => Debug.WriteLine(str);
+#if DEBUG || DEBUG_SPY_ALWAYS
```

Comment:
I feel like this should be...

```cs
#if DEBUG
    private static void DebugLogger(string str) => Debug.WriteLine(str); 
#elif DEBUG_SPY_ALWAYS
    private static void DebugLogger(string str) => NativeMethods.OutputDebugString(str); 
#endif
```

I.E. the native call only gets used for Release builds.

Reply:
I agree.